File: example/java/jimmer-sql-graphql/src/main/java/org/babyfish/jimmer/sql/example/graphql/cfg/SqlClientConfig.java
package org.babyfish.jimmer.sql.example.graphql.cfg;

import org.babyfish.jimmer.spring.repository.SpringConnectionManager;
import org.babyfish.jimmer.sql.DraftInterceptor;
import org.babyfish.jimmer.sql.JSqlClient;
import org.babyfish.jimmer.sql.cache.CacheFactory;
import org.babyfish.jimmer.sql.dialect.H2Dialect;
import org.babyfish.jimmer.sql.dialect.MySqlDialect;
import org.babyfish.jimmer.sql.example.graphql.entities.JimmerModule;
import org.babyfish.jimmer.sql.filter.Filter;
import org.babyfish.jimmer.sql.runtime.*;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;

import javax.sql.DataSource;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.Reader;
import java.sql.SQLException;
import java.util.List;

@Configuration
public class SqlClientConfig {

    @Bean
    public JSqlClient sqlClient(
            DataSource dataSource,
            @Value("${spring.datasource.url}") String jdbcUrl,
            List<DraftInterceptor<?>> interceptors,
            List<Filter<?>> filters,
            @Autowired(required = false) CacheFactory cacheFactory
    ) {
        boolean isH2 = jdbcUrl.startsWith("jdbc:h2:");
        JSqlClient sqlClient = JSqlClient.newBuilder()
                .setConnectionManager(new SpringConnectionManager(dataSource))
                .setEntityManager(JimmerModule.ENTITY_MANAGER)
                .setDialect(
                        isH2 ? new H2Dialect() : new MySqlDialect() // Support sequence
                )
                .setExecutor(Executor.log())
                .addDraftInterceptors(interceptors)
                .addFilters(filters)
                .setCaches(it -> {
                    if (cacheFactory != null) {
                        it.setCacheFactory(cacheFactory);
                    }
                })
                .build();
        if (isH2) {
            initializeH2Database(sqlClient);
        }
        return sqlClient;
    }

    private void initializeH2Database(JSqlClient sqlClient) {
        sqlClient.getConnectionManager().execute(con -> {
            InputStream inputStream = SqlClientConfig.class
                    .getClassLoader()
                    .getResourceAsStream("h2-database.sql");
            if (inputStream == null) {
                throw new RuntimeException("no h2-database.sql");
            }
            try (Reader reader = new InputStreamReader(inputStream)) {
                char[] buf = new char[1024];
                StringBuilder builder = new StringBuilder();
                while (true) {
                    int len = reader.read(buf);
                    if (len == -1) {
                        break;
                    }
                    builder.append(buf, 0, len);
                }
                con.createStatement().execute(builder.toString());
            } catch (IOException | SQLException ex) {
                throw new RuntimeException("Cannot initialize h2 database", ex);
            }
            return null;
        });
    }
}

File: example/kotlin/jimmer-sql-kt/src/main/kotlin/org/babyfish/jimmer/example/kt/sql/cfg/SqlClientConfig.kt
package org.babyfish.jimmer.example.kt.sql.cfg

import org.babyfish.jimmer.example.kt.sql.model.*
import org.babyfish.jimmer.sql.DraftInterceptor
import org.babyfish.jimmer.sql.cache.CacheFactory
import org.babyfish.jimmer.sql.dialect.H2Dialect
import org.babyfish.jimmer.sql.dialect.MySqlDialect
import org.babyfish.jimmer.sql.filter.Filter
import org.babyfish.jimmer.sql.kt.KSqlClient
import org.babyfish.jimmer.sql.kt.filter.KFilter
import org.babyfish.jimmer.sql.kt.newKSqlClient
import org.babyfish.jimmer.sql.runtime.EntityManager
import org.babyfish.jimmer.sql.runtime.ScalarProvider
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Value
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.jdbc.datasource.DataSourceUtils
import java.io.InputStreamReader
import java.sql.Connection
import javax.sql.DataSource

@Configuration
class SqlClientConfig {

    private val LOGGER = LoggerFactory.getLogger(SqlClientConfig::class.java)

    @Bean
    fun sqlClient(
        dataSource: DataSource,
        @Value("\${spring.datasource.url}") jdbcUrl: String,
        interceptors: List<DraftInterceptor<*>>,
        filters: List<KFilter<*>>,
        cacheFactory: CacheFactory? // Optional dependency
    ): KSqlClient {
        val isH2 = jdbcUrl.startsWith("jdbc:h2:")
        return newKSqlClient {

            setConnectionManager {
                /*
                 * It's very important to use
                 *      "org.springframework.jdbc.datasource.DataSourceUtils"!
                 * This is spring transaction aware ConnectionManager
                 */
                val con: Connection = DataSourceUtils.getConnection(dataSource)
                try {
                    proceed(con)
                } finally {
                    DataSourceUtils.releaseConnection(con, dataSource)
                }
            }

            setExecutor {
                /*
                 * Log SQL and variables
                 */
                LOGGER.info("Execute sql : \"{}\", with variables: {}", sql, variables)
                proceed()
            }

            setDialect(if (isH2) H2Dialect() else MySqlDialect())

            addDraftInterceptors(interceptors)

            addFilters(filters)

            setEntityManager(
                EntityManager(
                    BookStore::class.java,
                    Book::class.java,
                    Author::class.java,
                    TreeNode::class.java
                )
            )
            setCaches {
                if (cacheFactory != null) {
                    setCacheFactory(cacheFactory)
                }
            }
        }.also {
            if (isH2) {
                initializeH2Database(it)
            }
        }
    }

    private fun initializeH2Database(sqlClient: KSqlClient) {
        sqlClient.executeNativeSql { con: Connection ->
            val inputStream = SqlClientConfig::class.java
                .classLoader
                .getResourceAsStream("h2-database.sql") ?: throw RuntimeException("no h2-database.sql")
            val sql = InputStreamReader(inputStream).use { reader ->
                reader.readText()
            }
            con.createStatement().execute(sql)
        }
    }
}